File: ios/SxwnlCalendar/SxwnlCalendar/Models/BaziCalc.swift
import Foundation

// ════════════════════════════════════════════════════════════════
//  BaziCalc — 与鸿蒙 model/BaziInput.ets 对齐的纯计算工具
//
//  仅依赖 SxwnlBridge 提供的数据接口, 不引入任何 UI。
// ════════════════════════════════════════════════════════════════

enum BaziCalc {

    static let GAN: [String] = ["甲", "乙", "丙", "丁", "戊",
                                "己", "庚", "辛", "壬", "癸"]
    static let ZHI: [String] = ["子", "丑", "寅", "卯", "辰", "巳",
                                "午", "未", "申", "酉", "戌", "亥"]

    /// 六十甲子序号 → 名称
    static func jiaZiName(_ idx: Int) -> String {
        let i = ((idx % 60) + 60) % 60
        return GAN[i % 10] + ZHI[i % 12]
    }

    /// 干支索引 → 六十甲子序号
    static func gz60(gan: Int, zhi: Int) -> Int {
        for n in 0..<60 where n % 10 == gan && n % 12 == zhi {
            return n
        }
        return 0
    }

    /// 五虎遁: 由年干得 12 个合法月柱(寅月起)的六十甲子序号
    static func validMonths(yearIdx: Int) -> [Int] {
        let yg = ((yearIdx % 60) + 60) % 60 % 10
        let base = ((yg % 5) * 2 + 2) % 10
        return (0..<12).map { k in
            gz60(gan: (base + k) % 10, zhi: (2 + k) % 12)
        }
    }

    /// 五鼠遁: 由日干得 12 个合法时柱(子时起)的六十甲子序号
    static func validHours(dayIdx: Int) -> [Int] {
        let dg = ((dayIdx % 60) + 60) % 60 % 10
        let base = ((dg % 5) * 2) % 10
        return (0..<12).map { h in
            gz60(gan: (base + h) % 10, zhi: h)
        }
    }

    /// 在合法列表中按地支保留, 返回新的六十甲子序号
    static func remapByZhi(validList: [Int], oldIdx: Int) -> Int {
        let z = ((oldIdx % 60) + 60) % 60 % 12
        for v in validList where v % 12 == z { return v }
        return validList.first ?? 0
    }

    // ── 流年(立春界)干支 ──
    static func lnGan(_ year: Int) -> Int { (((year - 4) % 10) + 10) % 10 }
    static func lnZhi(_ year: Int) -> Int { (((year - 4) % 12) + 12) % 12 }

    // ── 年份解析 (B212 / *212 / -211) ──
    static func parseYear(_ s: String) -> Int? {
        let astro = YearUtil.yearStrToAstro(s)
        return astro == Int.min ? nil : astro
    }

    /// 天文年 → 显示文本: 2024 → "2024年", -210 → "公元前211年"
    static func formatYear(_ y: Int) -> String {
        if y <= 0 { return "公元前\(1 - y)年" }
        return "\(y)年"
    }

    // ── 月长 ──
    static func solarDaysInMonth(year: Int, month: Int) -> Int {
        if month == 2 {
            let leap = (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)
            return leap ? 29 : 28
        }
        if month == 4 || month == 6 || month == 9 || month == 11 { return 30 }
        return 31
    }

    static func lunarDaysInMonth(year: Int, month: Int,
                                 isLeap: Bool, isSpec: Bool) -> Int {
        let d = SxwnlBridge.getLunarMonthDays(year: year, month: month,
                                              isLeap: isLeap, isSpec: isSpec)
        return (d == 29 || d == 30) ? d : 30
    }

    static func pad2(_ n: Int) -> String {
        n < 10 ? "0\(n)" : "\(n)"
    }

    /// 出生时间一条记录文本: "公历 2026年06月02日 13点16分"
    static func formatRecord(_ sel: DateSelection,
                             lunarMonths: [LunarMonth]) -> String {
        let ymd = "\(formatYear(sel.year))\(pad2(sel.month))月\(pad2(sel.day))日"
        let hm = "\(sel.hour)点\(pad2(sel.minute))分"
        if sel.inputMode == .lunar {
            var mName = "\(sel.month)月"
            for m in lunarMonths where m.month == sel.month && m.isLeap == sel.isLeap {
                mName = m.name
                break
            }
            return "农历 \(formatYear(sel.year))\(mName)\(sel.day)日 \(hm)"
        }
        return "公历 \(ymd) \(hm)"
    }

    /// 表单 → SxwnlBridge.calcBazi 的入参 (调用时直接展开传入)
    struct BaziCallParams {
        let name: String
        let gender: Bool
        let year: Int
        let month: Int
        let day: Int
        let hour: Int
        let minute: Int
        let isLunar: Bool
        let isLeap: Bool
        let isSpec: Bool
        let astEnabled: Bool
        let longitude: Double
        let latitude: Double
        let lifa: Int
    }

    static func toParams(_ form: BaziFormData) -> BaziCallParams {
        let s = form.selection
        return BaziCallParams(
            name: form.name.isEmpty ? "匿名" : form.name,
            gender: form.gender,
            year: s.year, month: s.month, day: s.day,
            hour: s.hour, minute: s.minute,
            isLunar: s.inputMode == .lunar,
            isLeap: s.inputMode == .lunar ? s.isLeap : false,
            isSpec: s.inputMode == .lunar ? s.isSpec : false,
            astEnabled: form.astEnabled,
            longitude: form.longitude,
            latitude: form.latitude,
            lifa: form.lifaModern.rawValue
        )
    }
}

File: android/app/src/main/java/com/sxwnl/calendar/util/BaziCalc.kt
package com.sxwnl.calendar.util

import com.sxwnl.calendar.data.BaziParams
import com.sxwnl.calendar.data.LunarMonth

/**
 * 与鸿蒙端 model/BaziInput.ets 对齐的八字纯计算工具(无 UI 依赖)。
 */
object BaziCalc {

    val GAN: Array<String> = arrayOf(
        "甲", "乙", "丙", "丁", "戊", "己", "庚", "辛", "壬", "癸"
    )
    val ZHI: Array<String> = arrayOf(
        "子", "丑", "寅", "卯", "辰", "巳",
        "午", "未", "申", "酉", "戌", "亥"
    )

    /** 输入模式 */
    const val MODE_SOLAR = 0
    const val MODE_LUNAR = 1
    const val MODE_REVERSE = 2

    /** 输入模式 — 枚举形式 (与鸿蒙端 BaziInput.ets 的整型常量等价) */
    enum class BirthInputMode(val rawValue: Int) {
        SOLAR(MODE_SOLAR), LUNAR(MODE_LUNAR), REVERSE(MODE_REVERSE);

        companion object {
            fun of(v: Int): BirthInputMode = when (v) {
                MODE_LUNAR -> LUNAR
                MODE_REVERSE -> REVERSE
                else -> SOLAR
            }
        }
    }

    /** 出生时间选择结果 */
    data class DateSelection(
        val inputMode: BirthInputMode = BirthInputMode.SOLAR,
        val year: Int = 1990,
        val month: Int = 1,
        val day: Int = 1,
        val hour: Int = 12,
        val minute: Int = 0,
        val isLeap: Boolean = false,
        val isSpec: Boolean = false
    )

    /** 历法 — 仅 3 项 (与鸿蒙 LIFA_OPTIONS 一致) */
    const val LIFA_DING_QI = 11
    const val LIFA_PING_DONG_ZHI = 12
    const val LIFA_PING_XIA_ZHI = 13

    data class LiFaOption(val label: String, val value: Int)

    val LIFA_OPTIONS: List<LiFaOption> = listOf(
        LiFaOption("定气", LIFA_DING_QI),
        LiFaOption("平气定夏至", LIFA_PING_XIA_ZHI),
        LiFaOption("平气定冬至", LIFA_PING_DONG_ZHI)
    )

    fun jiaZiName(idx: Int): String {
        val i = ((idx % 60) + 60) % 60
        return GAN[i % 10] + ZHI[i % 12]
    }

    fun gz60(gan: Int, zhi: Int): Int {
        for (n in 0 until 60) {
            if (n % 10 == gan && n % 12 == zhi) return n
        }
        return 0
    }

    /** 五虎遁: 由年干得 12 个合法月柱(寅月起)的六十甲子序号 */
    fun validMonths(yearIdx: Int): IntArray {
        val yg = ((yearIdx % 60) + 60) % 60 % 10
        val base = ((yg % 5) * 2 + 2) % 10
        return IntArray(12) { k -> gz60((base + k) % 10, (2 + k) % 12) }
    }

    /** 五鼠遁: 由日干得 12 个合法时柱(子时起)的六十甲子序号 */
    fun validHours(dayIdx: Int): IntArray {
        val dg = ((dayIdx % 60) + 60) % 60 % 10
        val base = ((dg % 5) * 2) % 10
        return IntArray(12) { h -> gz60((base + h) % 10, h) }
    }

    fun remapByZhi(validList: IntArray, oldIdx: Int): Int {
        val z = ((oldIdx % 60) + 60) % 60 % 12
        for (v in validList) {
            if (v % 12 == z) return v
        }
        return validList.firstOrNull() ?: 0
    }

    fun lnGan(year: Int): Int = (((year - 4) % 10) + 10) % 10
    fun lnZhi(year: Int): Int = (((year - 4) % 12) + 12) % 12

    fun parseYear(s: String): Int? {
        val v = YearUtil.yearStrToAstro(s)
        return if (v == Int.MIN_VALUE) null else v
    }

    fun formatYear(y: Int): String = if (y <= 0) "公元前${1 - y}年" else "${y}年"

    fun solarDaysInMonth(year: Int, month: Int): Int = when (month) {
        2 -> {
            val leap = (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)
            if (leap) 29 else 28
        }
        4, 6, 9, 11 -> 30
        else -> 31
    }

    fun pad2(n: Int): String = if (n < 10) "0$n" else "$n"

    /**
     * 农历日名(初一../三十). 统一走底层 C++(复用 sx_lang_zh.h 的 Rmc[] 表),
     * 不在各平台重复维护字符串表; native 异常时回退 "${d}日".
     */
    fun lunarDayName(d: Int): String {
        val s = com.sxwnl.calendar.bridge.SxwnlNative.getLunarDayName(d)
        return s.ifEmpty { "${d}日" }
    }

    /** 出生时间一条记录文本 */
    fun formatRecord(p: BaziParams, lunarMonths: List<LunarMonth>): String {
        val ymd = "${formatYear(p.year)}${pad2(p.month)}月${pad2(p.day)}日"
        val hm = "${p.hour}点${pad2(p.minute)}分"
        if (p.inputMode == MODE_LUNAR) {
            val mName = lunarMonths.firstOrNull {
                it.month == p.month && it.isLeap == p.isLeap
            }?.name ?: "${p.month}月"
            return "农历 ${formatYear(p.year)}${mName}${lunarDayName(p.day)} $hm"
        }
        return "公历 $ymd $hm"
    }
}
